logstash-core/lib/logstash/pipeline.rb

```diff
@@ -45,11 +46,21 @@ module LogStash; class Pipeline
 
   def initialize(config_str, settings = LogStash::SETTINGS, namespaced_metric = nil)
     @config_str = config_str
+    @config_hash = Digest::SHA1.hexdigest(@config_str)
+    # Every time #plugin is invoked this is incremented to give each plugin
+    # a unique id when auto-generating plugin ids
+    @plugin_counter ||= 0 
+    
     @logger = Cabin::Channel.get(LogStash)
     @settings = settings
     @pipeline_id = @settings.get_value("pipeline.id") || self.object_id
     @reporter = LogStash::PipelineReporter.new(@logger, self)
 
+    # Maybe make this configurable via an option at some point
+    @output_delegator_strategy_registry = ::LogStash::OutputDelegatorStrategyRegistry.instance
```

Comment:
Good point. My original intent was to pass this as an argument to the output delegator. I think I'm going to go ahead and make that refactor now as its worth it. 

Ultimately this should simply be the default value, with the calling code able to override it so we can do richer testing.